cmake/crc32c.cmake

```diff
@@ -78,38 +78,41 @@ check_cxx_source_compiles_with_flags("${ARM_CRC_CXXFLAGS}" "
   " HAVE_ARM64_CRC32C
 )
 
+add_library(crc32c_common INTERFACE)
+target_compile_definitions(crc32c_common INTERFACE
+  HAVE_BUILTIN_PREFETCH=$<BOOL:${HAVE_BUILTIN_PREFETCH}>
+  HAVE_MM_PREFETCH=$<BOOL:${HAVE_MM_PREFETCH}>
+  HAVE_STRONG_GETAUXVAL=$<BOOL:${HAVE_STRONG_GETAUXVAL}>
+  BYTE_ORDER_BIG_ENDIAN=$<STREQUAL:${CMAKE_CXX_BYTE_ORDER},BIG_ENDIAN>
+)
+target_link_libraries(crc32c_common INTERFACE core_base_interface)
+
 add_library(crc32c STATIC EXCLUDE_FROM_ALL
   ${PROJECT_SOURCE_DIR}/src/crc32c/src/crc32c.cc
   ${PROJECT_SOURCE_DIR}/src/crc32c/src/crc32c_portable.cc
 )
-
-target_compile_definitions(crc32c
-  PRIVATE
-    HAVE_BUILTIN_PREFETCH=$<BOOL:${HAVE_BUILTIN_PREFETCH}>
-    HAVE_MM_PREFETCH=$<BOOL:${HAVE_MM_PREFETCH}>
-    HAVE_STRONG_GETAUXVAL=$<BOOL:${HAVE_STRONG_GETAUXVAL}>
-    HAVE_SSE42=$<BOOL:${HAVE_SSE42}>
-    HAVE_ARM64_CRC32C=$<BOOL:${HAVE_ARM64_CRC32C}>
-    BYTE_ORDER_BIG_ENDIAN=$<STREQUAL:${CMAKE_CXX_BYTE_ORDER},BIG_ENDIAN>
-)
-
 target_include_directories(crc32c
   PUBLIC
     $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/src/crc32c/include>
 )
+target_link_libraries(crc32c PRIVATE crc32c_common)
 
 if(HAVE_SSE42)
-  target_sources(crc32c PRIVATE ${PROJECT_SOURCE_DIR}/src/crc32c/src/crc32c_sse42.cc)
-  set_property(SOURCE ${PROJECT_SOURCE_DIR}/src/crc32c/src/crc32c_sse42.cc
-    APPEND PROPERTY COMPILE_OPTIONS ${SSE42_CXXFLAGS}
+  add_library(crc32c_sse42 STATIC EXCLUDE_FROM_ALL
```

Comment:
Couldn't these be OBJECT libraries since the static libs themselves are never useful?

Reply:
Yes, it could. But CMake's developers recommend to use `STATIC` libraries if there is no compelling reasons for `OBJECT` ones. For example, from [Professional CMake: A Practical Guide 18th Edition](https://crascit.com/professional-cmake/release-notes/18th-edition/)

> Prefer defining static libraries over object libraries. Static libraries are simpler, have more
complete and robust support from earlier CMake versions and they are well understood by most
developers. Object libraries have their uses, but they are also less flexible than static libraries.